File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:

    int searchNoDublicates(vector<int>& arr,int target){

        int start = 0 ;
        int end = arr.size() - 1;

        while(start <= end){

            int mid = (start + end)/2;

            if(arr[mid] == target)return mid;

            else if(arr[mid] >= arr[start]){ // the sorted part is left
                if(target >= arr[start] && target < arr[mid]){
                    end = mid - 1;
                }else{
                    start = mid + 1;
                }
            }
            else{
                if(target > arr[mid] && target <= arr[end]){
                    start = mid + 1;
                }else{
                    end = mid - 1;
                }
            }

        }   

        return -1;

    }



        
    int search(vector<int>& nums, int target) {
        
        // int start = 0 , end = nums.size() - 1;

        // while(start <= end){

        //     int mid = (start + end) / 2;

        //     if(nums[mid] == target){
        //         return mid ; 
        //     }

        //     else if(nums[mid] > nums[end]){
        //         if(target >= nums[start] && target < nums[mid]){
        //             end = mid - 1;
        //         } else{
        //             start = mid + 1;
        //         }
        //     }

        //     else{
        //         if(target > nums[mid] && target <= nums[end]){
        //             start = mid + 1 ;
        //         }else{
        //             end = mid - 1;
        //         }
                
        //     }

        //     // return -1;
            
        // }

        return searchNoDublicates(nums,target);

        
    
        
        
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int k) {
        int low = 0 ;
        int high = nums.length -1 ;

        while(low <= high){

            int mid = low + (high - low)/2 ;

            if(nums[mid] == k){
                return mid;
            }

            if(nums[mid] >= nums[low]) // left side sorted hai
            {
                if(k >= nums[low] && k < nums[mid]){
                    high = mid - 1;
                }
                else{
                    low = mid + 1 ;
                }
            }


            else{ // right part sorted hai isme check kaar lo element ko


                if(k > nums[mid] && k <= nums[high]){
                    low  = mid + 1 ;
                }else{
                    high = mid - 1 ;
                }

            }
        }

        return -1;
    }
}